### ISSUE TITLE
Comparison of Fixnum with nil failed (ArgumentError)

### ISSUE THREAD

**Issue by MauPalantir** [swipely/sumo#13]
I get this error when I try to use sumo-search either from CLI or as lib. It seems the error comes from collection.rb:

Stack trace:
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:38:in `<'
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:38:in `get_new_status'
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:32:in `status'
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:47:in `total'
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:68:in `limit'
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:62:in `has_results?'
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:57:in `page'
      ./vendor/cache/sumo-9220cb1f5cd3/lib/sumo/collection.rb:21:in `each'

My lib call was:
```ruby
search = Sumo.search(
  query: sumo_query, # Valid and has results.
  time_zone: 'UTC',
  from: from.strftime('%FT%T'),
  to: now.strftime('%FT%T')
)
```

**Comment by nahiluhmot** [comment]
Does the error occur when you create the search or try to paginate through the messages? If the `Sumo.search` call returns successfully, could you ensure that the search has an ID?

**Comment by MauPalantir** [comment]
there were two problems,
https://github.com/swipely/sumo/compare/master...MauPalantir:fixnum-comparison

The root cause was that Sumo api redirected my request to api endpoint api.us2.sumologic.com, and because of not getting an explicit error, the gem tried to paginate, but it did not get the results obviously.

(my solution is not good, its first draft, we should rather handle the redirection as an error or so)

**Comment by nahiluhmot** [comment]
Ah, that makes sense. I think that the best thing to do would be to add redirect-following middleware. It looks like `Excon` [has an optional builtin](https://github.com/excon/excon/blob/master/lib/excon/middlewares/redirect_follower.rb) for that, so we could probably use that. Alternatively, the client could explicitly handle `3xx` level HTTP errors. I'd be happy to accept a PR that does either.

Thanks for debugging!

**Comment by MauPalantir** [comment]
Theat wat the excon response:
```ruby
#<Excon::Response:0x007fc283bcd6b0 
@data={:body=>"{\n  \"status\" : 301,\n  \"id\" : \"ZVLMV-KPHD7-3PH1Q\",\n  \"code\" : \"moved\",\n  \"message\" : \"The requested resource SHOULD be accessed through returned URI in Location Header.\"\n}", 
:headers=>{"Content-Type"=>"application/json; charset=ISO-8859-1", "Date"=>"Tue, 15 Sep 2015 10:22:59 GMT", "Location"=>"https://api.us2.sumologic.com/api/v1/search/jobs", "Content-Length"=>"170", "Connection"=>"keep-alive"}, 
:status=>301, :status_line=>"HTTP/1.1 301 Moved Permanently\r\n", 
:reason_phrase=>"Moved Permanently", :remote_ip=>"50.17.203.153", 
:local_port=>55110, :local_address=>"192.168.1.210"}, 
@body="{\n  \"status\" : 301,\n  \"id\" : \"ZVLMV-KPHD7-3PH1Q\",\n  \"code\" : \"moved\",\n  \"message\" : \"The requested resource SHOULD be accessed through returned URI in Location Header.\"\n}", @headers={"Content-Type"=>"application/json; charset=ISO-8859-1", "Date"=>"Tue, 15 Sep 2015 10:22:59 GMT", "Location"=>"https://api.us2.sumologic.com/api/v1/search/jobs", "Content-Length"=>"170", "Connection"=>"keep-alive"}, @status=301, @remote_ip="50.17.203.153", @local_port=55110, @local_address="192.168.1.210">
```

And the contents of state hash
```
{:get_values=>#<Proc:0x007fc283bcc800@/Users/Mau/Development/sumo/lib/sumo/search.rb:33>, :get_status=>#<Proc:0x007fc283bcc7d8@/Users/Mau/Development/sumo/lib/sumo/search.rb:34>, :count_key=>"messageCount"}
```

Okay, gonna implement a simple middleware. Should I keep the state keys’ validation in place?

**Comment by nahiluhmot** [comment]
Let's remove it for now, the validation really should happen in the client.

**Comment by MauPalantir** [comment]
Hmm, I removed all but enabling the middleware, but I get a 401 then.

```
#<Excon::Response:0x007f900344bd40 @data={:body=>"<html>\n<head>\n<meta http-equiv=\"Content-Type\" content=\"text/html; charset=ISO-8859-1\"/>\n<title>Error 401 Full authentication is required to access this resource</title>\n</head>\n<body><h2>HTTP ERROR 401</h2>\n<p>Problem accessing /api/v1/search/jobs. Reason:\n<pre>    Full authentication is required to access this resource</pre></p>[...]</body>\n</html>\n", :headers=>{"Cache-Control"=>"must-revalidate,no-cache,no-store", "Cache-control"=>"must-revalidate,no-cache,no-store, no-cache=\"set-cookie\"", "Content-Type"=>"text/html; charset=iso-8859-1", "Date"=>"Tue, 15 Sep 2015 16:02:50 GMT", "Set-Cookie"=>"JSESSIONID=detdns4lhzx8f5cal6ky7qzd;Path=/api, AWSELB=0989F5E71E23D67F079F05E90403677CFCB21C735E635CC2E591ED1CFC524DA25957DCDC1905B9402105C7286036BF9AB7CE305BE8B8D82908B66AEBAE687BAA618DB12638;PATH=/", "WWW-Authenticate"=>"Basic realm=\"Sumo\"", "Content-Length"=>"1475", "Connection"=>"keep-alive"}, :status=>401, :status_line=>"HTTP/1.1 401 Full authentication is required to access this resource\r\n", :reason_phrase=>"Full authentication is required to access this resource", :remote_ip=>"54.214.26.68", :local_port=>58049, :local_address=>"192.168.1.210"}, @body="<html>\n<head>\n<meta http-equiv=\"Content-Type\" content=\"text/html; charset=ISO-8859-1\"/>\n<title>Error 401 Full authentication is required to access this resource</title>\n</head>\n<body><h2>HTTP ERROR 401</h2>\n<p>Problem accessing /api/v1/search/jobs. Reason:\n<pre>    Full authentication is required to access this resource</pre></p>[...]</body>\n</html>\n", @headers={"Cache-Control"=>"must-revalidate,no-cache,no-store", "Cache-control"=>"must-revalidate,no-cache,no-store, no-cache=\"set-cookie\"", "Content-Type"=>"text/html; charset=iso-8859-1", "Date"=>"Tue, 15 Sep 2015 16:02:50 GMT", "Set-Cookie"=>"JSESSIONID=detdns4lhzx8f5cal6ky7qzd;Path=/api, AWSELB=0989F5E71E23D67F079F05E90403677CFCB21C735E635CC2E591ED1CFC524DA25957DCDC1905B9402105C7286036BF9AB7CE305BE8B8D82908B66AEBAE687BAA618DB12638;PATH=/", "WWW-Authenticate"=>"Basic realm=\"Sumo\"", "Content-Length"=>"1475", "Connection"=>"keep-alive"}, @status=401, @remote_ip="54.214.26.68", @local_port=58049, @local_address="192.168.1.210">
```

**Comment by MauPalantir** [comment]
hmm, it seems to delete the Authorization header https://github.com/excon/excon/blob/master/lib/excon/middlewares/redirect_follower.rb#L29

**Comment by nahiluhmot** [comment]
What a pain! How about manually following redirects in `Sumo::Client` then? Maybe change request to something like:

```ruby
def request(hash, &block)
  response = connection.request(add_defaults(hash), &block)
  return follow_redirect!(response) if redirect?(response)
  ...
end
```

**Comment by MauPalantir** [comment]
yeah, Im working on this, and the redirection itself works in 2490f5a94e3799c978422fb9c2d8bd21b4c433d7, though now I have to dig into search.rb, since the redirect JSON adds a "request id" or so, which overrides the original job ID. Thus I get an internal server error after redirection:D

**Comment by MauPalantir** [comment]
Ok, it was only a bug in the redirection process :relieved: 
Created the PR at 
https://github.com/swipely/sumo/pull/14

**Comment by nahiluhmot** [comment]
Awesome! I'll try to review it today, but if I don't have time it'll certainly be tomorrow.

**Comment by MauPalantir** [comment]
ping?

**Comment by nahiluhmot** [comment]
Fixed in #14.
